Positron Notebooks: Find

seeM opened the pull request:
Addresses #10064.

I added a notebook editor extension system based on the upstream ones for Monaco and Notebooks. Looked like a useful way to separate out components like this.

There were a bunch of existing find widget implementations, but unfortunately they all differ in subtle ways that made them unusable in our case. I made a generic React-based find widget that we can use in other places too like the Console. The idea is to also extend that widget with replace functionality when we get there.

I started looking into unit/integration tests but it got complicated so will do that separately.

_Note: I used Claude pretty heavily for this PR, but I kept a heavy hand on the architecture throughout._

### Release Notes

#### New Features

- N/A

#### Bug Fixes

- N/A

### QA Notes

@:positron-notebooks @:web

seeM committed:
positron notebook contributions

seeM committed:
fix ctor type

seeM committed:
get contribution from notebook

seeM committed:
sketch outline of notebook find contribution

seeM committed:
lift state

seeM committed:
close widget on esc

seeM committed:
find widget visible context key; esc to close

seeM committed:
fix extension services types

seeM committed:
wrap widget to accept observables; attempt search and text editor decorations

seeM committed:
fix: clear old decorations

seeM committed:
fix toggles

seeM committed:
wip: match count formatting

seeM committed:
move find widget out of notebook container so that keys like `a` dont run notebook actions

seeM committed:
use code oss find input widget

seeM committed:
renames

seeM committed:
clean css

seeM committed:
inline find input component

seeM committed:
extract it again lol

seeM committed:
remove unneeded container

seeM committed:
add find results; remove separate additionalToggles prop

seeM committed:
left align find results

seeM committed:
oops wrong css

seeM committed:
disable prev/next when no matches

seeM committed:
disable up/down buttons when no matches

seeM committed:
init goto prev/next

seeM committed:
find from cursor

seeM committed:
focus input context key

seeM committed:
attempt enter/shift+enter actions

seeM committed:
trying to fix enter in find widget by using a shared context key service scoped to an off screen element

seeM committed:
Revert "trying to fix enter in find widget by using a shared context key service scoped to an off screen element"

This reverts commit 4f82db6781878d8a046e278c0b983e80e266bb2c.

seeM committed:
just handle enter and shift enter manually...

seeM committed:
improve focus on nav to match

seeM committed:
exit on escape in find input

seeM committed:
cmd+f while editing a cell

seeM committed:
fix find decorations

seeM committed:
clear decorations when closing the find widget

seeM committed:
attempt to fix losing focus when find next goes to a new cell

seeM committed:
research when the find widget is shown

seeM committed:
decorate current match

seeM committed:
fix shift enter for find prev

seeM committed:
use current match in find next/prev

seeM committed:
extract renderer creation

seeM committed:
extract find instance

seeM committed:
lift find input options

seeM committed:
handle visibility and focus via observables

seeM committed:
minor cleanup

seeM committed:
move find widget into notebook container from its parent

seeM committed:
fix recursion error

seeM committed:
remove find in selection toggle

seeM committed:
stay in zero based indexing

seeM committed:
extract getCurrentCursor

midleman committed:
e2e test template

seeM committed:
cleaning up: introduce cell range, position, match helpers; introduce cell find decorations helper

seeM committed:
extract decorations to separate file

seeM committed:
localize user-facing strings

seeM committed:
extract position and range to separate files

seeM committed:
move getActiveEditorPosition to notebook instance

seeM committed:
polish

seeM committed:
todo comments

seeM committed:
stop using modal react renderer; animate & hide find widget

seeM committed:
cleanup classes and fix sliding off screen

seeM committed:
matching code oss style

seeM committed:
use react renderer

seeM committed:
fix first show animation

seeM committed:
remove animations for now

seeM committed:
store handle refs and query notebook

seeM committed:
research on show after hide

seeM committed:
research on notebook content changes; debounce search string and content changes

seeM committed:
dont debounce search query; reduce debounce of notebook content changes; reveal cell if outside viewport

seeM committed:
fix navigating from rendered markdown cell

seeM commented on src/vs/workbench/contrib/positronNotebook/browser/positronNotebook.contribution.ts:
```diff
@@ -343,7 +347,10 @@ registerAction2(class extends NotebookAction2 {
 			title: localize2('positronNotebook.selectUp', "Move Focus Up"),
 			keybinding: {
-				when: POSITRON_NOTEBOOK_EDITOR_CONTAINER_FOCUSED,
+				when: ContextKeyExpr.and(
```
Is this fine? Do we want to create another helper context key?

seeM commented on src/vs/workbench/contrib/positronNotebook/common/editor/position.ts:
```diff
@@ -0,0 +1,50 @@
+/*---------------------------------------------------------------------------------------------
```
Some helpers that wrap the existing position/range classes with a `cellIndex` as well

nstrayer reviewed (commented):
Code looks and architecture looks pretty good. I think an extension system will come in handy and feels reasonable to do our own rather than try and make the existing notebooks one work somehow. 

I noticed the find widgets drop shadow looks funky with our breadcrumb bar because there's no border. It seems like with plain editors the find widget triggers a dropshadow/border on the breadcrumb bar when the find widget is active / the user has scrolled. We should probably add this but it's not relevant to this PR really. 

<img width="334" height="164" alt="image" src="https://github.com/user-attachments/assets/a27954aa-bfed-451c-bcb7-6041d444eef5" />

seeM committed:
add parent container of notebook and contributions so that contributions can live through clear/setinput when changing active editor

seeM committed:
create a single scoped context key service for the editor

seeM committed:
reset visibility and focus observables on unmount

seeM committed:
focus input on show e.g. from notebook

seeM committed:
set initial context key values

dhruvisompura reviewed (commented):
You are probably already be aware of this but the top of the find widget is getting cut off.

<img width="405" height="149" alt="Image" src="https://github.com/user-attachments/assets/685562f9-3456-4b34-a6b4-118a112b9384" />

I also ended up in this situation where I had multiple find widgets when I was switching between files. My repro steps are:
1. Open Notebook A and Notebook B
2. In Notebook A, press `cmd+f`, and then type something in the search field
3. Switch to Notebook B and press `cmd+f`
4. Observe that another find widget appears

https://github.com/user-attachments/assets/d3bf4439-5aab-4e01-93ea-662e4c6afb6d

Somewhat related to the above issue, how can a user re-trigger the search to happen that's in the find widget when they switch notebooks? In VS Code, it looks like the find widget will re-trigger the search when you switch editor tabs and the search results get updated.

seeM committed:
fix padding

seeM committed:
align style with code oss simple find widget

seeM committed:
make widget wider

seeM committed:
remove unneeded row div

seeM committed:
show scroll decoration when find widget is visible

seeM committed:
find input history

seeM committed:
show history keybinding hint

seeM committed:
fix find widget history lost on hide

seeM committed:
store find widget history in storage

seeM committed:
rename contributions container to overlay container

seeM committed:
fix find widget not hiding when changing to another notebook tab

seeM committed:
rename cell.editor to cell.currentEditor

seeM committed:
add public `cell.editor` observable

seeM committed:
tidy

seeM committed:
attempt to fix decorations when remounting notebook

seeM committed:
trace logging; fix decorations on remount

seeM committed:
only attach editor to cell when model is set

seeM committed:
fix: stay in edit mode when showing find widget

seeM committed:
tidy: getOrSet

seeM committed:
fulfil icellviewmodel decorations methods; extract cell decoration manager

seeM committed:
revert attaching editor only after model is set

seeM committed:
nit

seeM committed:
re-add container value getter under new name

seeM committed:
tidy

seeM commented:
Thanks for the reviews! @nstrayer I implemented the observer for the scroll decoration. @dhruvisompura I fixed the issues with the find widget being cut off and duplicates when switching editors. Please take another look

seeM committed:
fix merge conflicts

seeM committed:
fix focus when empty

seeM committed:
fix infinite render loop

seeM committed:
fix run cell not showing

seeM commented:
Hmm struggling with some more focus issues! Will try again tomorrow

seeM committed:
fix notebook find stealing focus from output pane

seeM committed:
use notebook's context key service

nstrayer committed:
Fix keyboard shortcut bug by reverting changes made in diff view pr

seeM committed:
scope context keys to editor container

seeM committed:
dont manually override notebook editor focus context

seeM committed:
rework notebook context

seeM committed:
scoped context key service is always defined

seeM committed:
fix after rename

seeM committed:
handle all find widget keys via actions

seeM commented on src/vs/workbench/contrib/positronNotebook/browser/positronNotebook.contribution.ts:
```diff
@@ -595,6 +596,9 @@ registerAction2(class extends NotebookAction2 {
 			cell.insertCodeCellBelow();
+		} else {
+			// Empty notebook: add a code cell
+			notebook.addCell(CellKind.Code, 0, false);
```
Minor bug fix

seeM commented on src/vs/workbench/contrib/positronNotebook/browser/PositronNotebookEditor.tsx:
```diff
@@ -198,21 +217,32 @@ export class PositronNotebookEditor extends AbstractEditorWithViewState<INoteboo
-		parent.appendChild(this._parentDiv);
-		this._parentDiv.style.display = 'relative';
+
+		// Create the top-level editor container
```
@dhruvisompura @nstrayer another big'ish change is to introduce one more nesting to the DOM tree:

- `positron-notebook-editor` is what the context key service is scoped to and where we track focus, it's the parent node
    - `positron-notebook-container` is the React root for the notebook component - the fact that this is a React root is why I added another nesting, since this gets destroyed when the notebook instance is detached e.g. when changing the active editor
    - `positron-notebook-overlay-container` is a parent that overlay widgets (like the find widget) can append their own React roots to

dhruvisompura commented on src/vs/workbench/contrib/positronNotebook/browser/ContextKeysManager.ts:
```diff
@@ -4,23 +4,24 @@
+ * Context key that is set when the Positron notebook editor container is focused.
  */
-export const POSITRON_NOTEBOOK_EDITOR_CONTAINER_FOCUSED = new RawContextKey<boolean>('positronNotebookEditorContainerFocused', false);
+export const POSITRON_NOTEBOOK_EDITOR_FOCUSED = new RawContextKey<boolean>('positronNotebookEditorFocused', false, localize('positronNotebookFocused', "Whether a Positron notebook editor or a notebook editor widget (e.g. a cell editor or the find widget) has focus"));
```
I really like the change to have this context key be true anytime the notebook has focus! This behavior is a lot closer to how it works in other panes/editors.

dhruvisompura commented on src/vs/workbench/contrib/positronNotebook/browser/ContextKeysManager.ts:
```diff
@@ -142,50 +138,30 @@ export class PositronNotebookContextKeyManager extends Disposable {
-	setContainerFocused(focused: boolean): void {
-		this.positronEditorFocus?.set(focused);
-	}
-
```
Ahh, so removing this bit of code is what ensures `POSITRON_NOTEBOOK_EDITOR_FOCUSED` stays `true` when a user is editing cells.
